packages/core/src/odata/common/filter/filter-function.ts

```diff
@@ -1,97 +1,62 @@
 /* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
 
-import { EdmTypeShared } from '../edm-types';
-import { EntityBase, ODataVersionOf } from '../entity';
-import { Field, FieldType } from '../selectable';
-import { Filter } from './filter';
+import { MapType } from '@sap-cloud-sdk/util';
+import { EntityBase } from '../entity';
+import { BooleanFilterFunction } from './boolean-filter-function';
+import { NumberFilterFunction } from './number-filter-function';
+import { StringFilterFunction } from './string-filter-function';
+import { FilterFunctionParameterType } from './filter-function-base';
+
+export function filterFunction<EntityT extends EntityBase>(
+  functionName: string,
+  returnType: 'boolean' | 'bool',
+  ...parameters: FilterFunctionParameterType<EntityT>[]
+): BooleanFilterFunction<EntityT>;
+
+export function filterFunction<EntityT extends EntityBase>(
+  functionName: string,
+  returnType: 'int' | 'double' | 'decimal',
+  ...parameters: FilterFunctionParameterType<EntityT>[]
+): NumberFilterFunction<EntityT>;
+
+export function filterFunction<EntityT extends EntityBase>(
+  functionName: string,
+  returnType: 'string',
+  ...parameters: FilterFunctionParameterType<EntityT>[]
+): StringFilterFunction<EntityT>;
 
 /**
- * Data structure to represent OData V2 filter functions.
- * Use the factory function [[filterFunction]] to create instances of `FilterFunction`.
+ * Build a custom filter function.
+ * @param functionName - the name of the function, e.g., `substring`
+ * @param returnType - the return type of the filter function
+ * @param parameters - the parameter(s) used in the function
+ * @returns An instance of filter function suited for the given return type
  */
-export abstract class FilterFunction<
-  EntityT extends EntityBase,
-  ReturnT extends FieldType
-> {
-  /**
-   * EdmType of the return type of the filter function.
-   */
-  public edmType: EdmTypeShared<ODataVersionOf<EntityT>>;
-
-  /**
-   * Creates an instance of FilterFunction.
-   * @param functionName - Name of the function
-   * @param parameters - Representation of the parameters passed to the filter function
-   */
-  constructor(
-    public functionName: string,
-    public parameters: FilterFunctionParameterType<EntityT>[]
-  ) {}
-
-  /**
-   * @deprecated Since v1.21.0. There will be no replacement. Let us know if you were using this functionality.
-   * Serializes the filter function into a string
-   * @param parentFieldNames - Names of parents in case the function is part of a filter on a navigation property
-   * @returns The filter function as string
-   */
-  toString(parentFieldNames: string[] = []): string {
-    const params = this.parameters
-      .map(param => this.transformParameter(param, parentFieldNames))
-      .join(', ');
-    return `${this.functionName}(${params})`;
+export function filterFunction<EntityT extends EntityBase>(
```

Comment:
I would call this method a bit different to indicate, that it builds, construcuts filterfunction but is not itself a filter function.

Reply:
Phew, this function has been there for a while now. I think as it is supposed to also be used by our users the name is quite good. I would generally like to keep it, but independent of that this is out of scope for this PR is guess ;)